Extract java-base from java plugin into separate project; refactor consistent resolution and compile task convention configuration (#26379)
gradle/gradle: subprojects/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_8.5]]
== Upgrading from 8.4 and earlier

=== Potential breaking changes

==== Changes to Groovy task conventions

> The `groovy-base` plugin is now responsible for configuring source and target compatibility version conventions on any `GroovyCompile` tasks.

Reviewer: ```suggestion
The `groovy-base` plugin is now responsible for configuring source and target compatibility version conventions on all `GroovyCompile` tasks.
Previously, the `java-base` plugin configured these conventions for `GroovyCompile` tasks. 
```

> If you are using this task *without applying this plugin*, you will have to manually set compatibility versions on these tasks.
> We recommend applying the `groovy-base` plugin whenever working with Groovy language tasks.

Reviewer: ```suggestion
If you are using this task *without applying `groovy-base`*, you will have to manually set compatibility versions on these tasks.
In general, the `groovy-base` plugin should be applied when working with Groovy language tasks.
```

. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_8.5]]
== Upgrading from 8.4 and earlier

=== Potential breaking changes

> ==== Changes to `JavaBasePlugin` conventions

Reviewer 2: Like we talked about... these should be user facing.  We don't need to include details about internals if they don't impact most users

> The `java-base` plugin now assumes a maximum of one ` `link:{javadocPath}/org/gradle/jvm/component/internal/JvmSoftwareComponentInternal.html[JvmSoftwareComponentInternal]` component will ever be created.

Reviewer 3: ```suggestion
The `java-base` plugin now assumes a maximum of one `link:{javadocPath}/org/gradle/jvm/component/internal/JvmSoftwareComponentInternal.html[JvmSoftwareComponentInternal]` component will ever be created.
```

The plugin's `java` extension block will now apply source and target compatibility versions and consistent resolution settings to every such component.
Previously, the plugin would only apply these settings to component named `java`.

> As there is no public API for creating components of this type, we do not expect this change to affect many builds.

Reviewer: How could a user get into the state where this affects behavior? I cannot think of a case unless the user literally is writing:

`project.components.add(project.objects.newInstance(DefaultJvmSoftwareComponent,...))`

This is not a case we support and we should not be catering out upgrade guide towards blatant misuse of internal APIs like this
